### DISCUSSION TITLE
Handling 404 as an error on a page level

### DISCUSSION THREAD

**Discussion by codinginflow** [vercel/next.js#55138] [App Router] [answered]
When I request a non-existing resource from my (Express) backend, it returns a 404 response and my Next.js app throws this as an error.

Is below the correct approach to handle this on a page level? It's a bit awkward to wrap the UI into try/catch.

```
export default async function Page({ params: { slug } }: PageProps) {
  try {
    const { ... } = await getPost(slug);

    [...]

    return (
      <div className={styles.container}
        [...]
      </div>
    );
  } catch (error) {
    if (error instanceof NotFoundError) {
      notFound();
    } else {
      throw error; // Let error.tsx handle it
    }
  }
}
```

**Edit:** Would this be better?

```
export default async function Page({ params: { slug } }: PageProps) {
let post: BlogPost;

try {
  post = await getPost(slug);
} catch {
  if (error instanceof NotFoundError) {
    notFound();
  } else {
    throw error;
  }
}

return ... (UI)
```

**Comment by icyJoseph** [comment, marked as answer]
I guess the second approach is good, but you could also, maybe, refactor getPost or wrap it in some code that, returns null when theres no post?

```js
const getPost = async (slug) => {
// maybe wrap all of this with a try/catch
  const response = await someAPI(slug)
  if (!response.ok) {
   // maybe report or log
   return null
  }
  const post = await response.json()
  
  return post
}
```

And then you just check:

```js
export default async function Page({ params: { slug } }: PageProps) {
  const post = await getPost()

   if (!post) notFound();

   return <UI />
}
```

**Reply by codinginflow** [reply]
I think that's a good idea, thank you.
My call is wrapped into `React.cache` because I need to make it in `generateMetadata` as well. Does this look correct to you?

```
const getPost = cache(async (slug: string) => {
  try {
    return await BlogApi.getBlogPostBySlug(slug);
  } catch (error) {
    if (error instanceof NotFoundError) {
      return null;
    } else {
      throw error;
    }
  }
});
```

**Reply by icyJoseph** [reply]
Yeah I guess so, I am not 100% sure how cache handles errors, but intuitively it looks right.

**Comment by codinginflow** [comment]
Thank you. Turns out, I can call `notFound()` directly inside this function. This used to not work because `generateMetadata` couldn't handle errors (it would lead to the `global-error.tsx` instead of `error.tsx`) but this seems to work now.

**Reply by icyJoseph** [reply]
Right, `notFound` is sort of a falling house of cards, you call it, and there's an error boundary wrapping the rendering process, that checks if the thrown error is a NOT_FOUND error, and if so handles it appropriately. 

In other words, both `notFound` and `redirect` have to be called, as you render.

**Reply by icyJoseph** [reply]
Before my summer break I think I explored this a bit more here: https://github.com/vercel/next.js/discussions/53845#discussioncomment-6712344

**Reply by codinginflow** [reply]
Thank you. Errors thrown from `generateMetadata` used to not lead to the `error.tsx`, but the `global-error.tsx`. And since `generateMetadata` is called before the page is rendered (from my observation), errors would lead to the `global-error.tsx` (which isn't great because it's not wrapped into the `layout.tsx`.

But it seems that they fixed this.

So to clarify, calling `notFound()` both inside the page and inside `generateMetadata` is fine?

**Reply by icyJoseph** [reply]
It looks like that's the case: https://github.com/vercel/next.js/blame/12252c271140657164f34290a35a07f0a2d8d279/packages/next/src/lib/metadata/metadata.tsx#L81 :

```ts
      // If the error triggers in initial metadata resolving, re-resolve with proper error type.
      // They'll be saved for flight data, when hydrates, it will replaces the SSR'd metadata with this.
      // for not-found error: resolve not-found metadata
      if (!errorType && isNotFoundError(resolvedError)) {
```

**Reply by codinginflow** [reply]
Awesome, thank you very much!
